### ISSUE TITLE
Implement playbooks to deploy an Elasticsearch 7 cluster with xpack security options

### ISSUE THREAD

**Issue by manuasir** [wazuh/wazuh-ansible#189]
Hi team,

It'd be necessary to implement the required tasks and roles for deploying an Elasticsearch cluster with `xpack security` and SSL enabled.

- [x] Add the ability to set up `xpack` security options for Elastic 7.
- [x] Add the ability to generate certificated and distribute the `CA` between the nodes.
- [x] Secure Filebeat with certificates signed by the same CA.
- [x] Secure Kibana using users.

This issue will be edited in order to add more details.

Cheers

**Comment by manuasir** [comment]
Hi team,

This task has been started approaching these main points:

- Modify the Elasticsearch role and add variables for being able to enable XPACK security and SSL protocol.
- Modify the Kibana role for being able to enable SSL and security.
- Set the necessary variables for being able to generate certificates and a CA (only **one** node of the cluster should do this).
- Modify the `elasticsearch.yml` template so the options can be added.

The template should include the logic for the previous work. The tricky part is that a `instances.yml` file is required for generating the certificates, and both IP and hostnames should of each cluster master node have to be known previously. Probably Ansible `Facts` will be used so we can get the hostname.

The next steps will be the implementation of the distribution of the certificates and CA from one node to the others. Probably the Ansible [synchronization](https://docs.ansible.com/ansible/latest/modules/synchronize_module.html) method will be used for approaching this.

Cheers

**Comment by manuasir** [comment]
#### Updated 06/29/2019 CEST

The development of this task is being done in [this branch](https://github.com/wazuh/wazuh-ansible/tree/elk-security).

- New tasks were added to implement the logic of the certificates generation:

```
- name: Write the instances.yml file in the selected node
  instances_file:
    src: instances.yml.j2
    dest: "/usr/share/elasticsearch/instances.yml"
  tags:
    - config
  when: 
    - node_generate_certs
    - elasticsearch_xpack_security
```

This is the first approach of the template:

```

# {{ ansible_managed }}

{% if node_generate_certs %}
instances:
{% for item in elasticsearch_cluster_nodes %}
    ip: {{ item.ip }}
    - "{{ item.name }}"
{% endfor %}
{% endif %}
```

Also, it's necessary to implement a way to detect that the file _already_ exists or not:
```
- name: Write the instances.yml file in the selected node
  instances_file:
    src: instances.yml.j2
    dest: "/usr/share/elasticsearch/instances.yml"
  tags:
    - config
  when: 
    - node_generate_certs
    - elasticsearch_xpack_security
```

I'm implementing this focusing on *variables*. This is the example of playbook I'm using for testing:

```
- hosts: 172.16.0.161
  roles:
    - {role: ../roles/elastic-stack/ansible-elasticsearch, elasticsearch_network_host: '172.16.0.161', node_generate_certs: true, elasticsearch_bootstrap_node: true, elasticsearch_cluster_nodes: ['172.16.0.162','172.16.0.163','172.16.0.161']}

- hosts: 172.16.0.162
  roles:
    - {role: ../roles/elastic-stack/ansible-elasticsearch, elasticsearch_network_host: '172.16.0.162', elasticsearch_master_candidate: true, elasticsearch_cluster_nodes: ['172.16.0.162','172.16.0.163','172.16.0.161']}

- hosts: 172.16.0.163
  roles:
    - {role: ../roles/elastic-stack/ansible-elasticsearch, elasticsearch_network_host: '172.16.0.163', elasticsearch_master_candidate: true, elasticsearch_cluster_nodes: ['172.16.0.162','172.16.0.163','172.16.0.161']}
```

Notice the `node_generate_certs` in the bootstrap node. The certificates and the `instances.yml` should only be generated in there.

The next steps would be the propagation of the certificates over the rest of the instances, probably not a hard part. The most tricky tasks will be the catch of the hostnames and IPs of each instance, both required for the `instances.yml` file. I'll keep this issue updated and will tell more details about the progression of the task. Also, don't hesitate to suggest any new idea.

Cheers

**Comment by jm404** [comment]
Hello team!

Today tasks I achieved are:

## Import certificates on generator node

Certificates generator node now moves his own certificates to the proper path. Also copies the `CA` certificate.

https://github.com/wazuh/wazuh-ansible/blob/5d8286f0deb525baa29673db1c0a7e9c95578bb3/roles/elastic-stack/ansible-elasticsearch/tasks/main.yml#L146-L151

Commit: [#5d8286f](https://github.com/wazuh/wazuh-ansible/commit/5d8286f0deb525baa29673db1c0a7e9c95578bb3)

## Elasticsearch XPack configuration

### Elasticsearch template

ES template has been updated and now incorpores the XPack Features:

https://github.com/wazuh/wazuh-ansible/blob/bd32839f27db43749bc3473a552db3b5828137fa/roles/elastic-stack/ansible-elasticsearch/templates/elasticsearch.yml.j2#L28-L41

Check :[#bd32839](https://github.com/wazuh/wazuh-ansible/commit/bd32839f27db43749bc3473a552db3b5828137fa)

### Default ES bootstrap password 

Default bootstraping password is now set by using the `elasticsearch-keystore` binary

https://github.com/wazuh/wazuh-ansible/blob/fdb1113a193c34a1537175e6ed92924f7a9385d6/roles/elastic-stack/ansible-elasticsearch/tasks/main.yml#L153-L157

Check: [#fdb1113](https://github.com/wazuh/wazuh-ansible/commit/fdb1113a193c34a1537175e6ed92924f7a9385d6)

### Certificates .zip file is now removed

A `rm` task via shell has been added to remove `certs.zip` file. 

Check: [#ab8cdd1](https://github.com/wazuh/wazuh-ansible/commit/ab8cdd13c63500a369f25139fe377b41f41a68a8)

## Filebeat XPack Security

Filebeat template and attributes have been upgraded to implement XPack attributes in the `filebeat.yml` file.

You can review such changes in: [#3ff5a19](https://github.com/wazuh/wazuh-ansible/commit/3ff5a194df017f6153b3fa3c1f47f863f16a5432) & [#6cffed9](https://github.com/wazuh/wazuh-ansible/commit/6cffed9218b78721edcb0e95bce1944b0227e09e)

## Kibana XPack Security 

Kibana installation templates and task files have been also upgraded following same methodology that previous components. The template and tasks files have been 

Check: [#2b18745](https://github.com/wazuh/wazuh-ansible/commit/2b18745cd0f32d457aa3ed8d535e40ecfa922bd4) & [#2656d89](https://github.com/wazuh/wazuh-ansible/commit/2656d89933d398e577fa30d3a9675e59eb4aa333)

Best regards,

Jose

**Comment by jm404** [comment]
Hi team,

Issue tasks have been finished. Changes and installation process are described below:

## Elastic XPack Security

### Objective

The objective is to deploy instances (can be Elasticsearch, Kibana or Manager+Filebeat) that implements and uses XPack Security features to encrypt communications using SSL/TLS.

Steps to configure such instances are:

### `elasticsearch/kibana/filebeat_xpack_security`

On every playbook, must be declared to true in order to enable the configuration of the certificates and the configuration of the templates . For example:

https://github.com/wazuh/wazuh-ansible/blob/2c14392e74cb3b1b85d5c3c1a7cf69e5ea36c69d/playbooks/wazuh-elastic_stack-distributed.yml#L36

### Create `instances.yml` file

Said file will save information about the instances we want to configure. The `name` is a must, other fields like `ip` or `dns` can be declared.

In our code, that file is generated using the template `instances.yml.j2` which iterates over the `instances` variable that must be declared in the playbook. In `wazuh-elastic_stack-distributed.yml` you can find an example:

https://github.com/wazuh/wazuh-ansible/blob/2c14392e74cb3b1b85d5c3c1a7cf69e5ea36c69d/playbooks/wazuh-elastic_stack-distributed.yml#L20-L30

### Generator node

The variable **node_certs_generator: true** designates  which node will be the one that will generate the certificates an unzip them.

**IMPORTANT NOTE**: `unzip` or `gtar` are required in such node.

Certificates to authenticate nodes are generated using:

-  `/usr/share/elasticsearch/bin/elasticsearch-certutil cert ca --pem --in {{node_certs_source}}/instances.yml --out {{node_certs_source}}/certs.zip` 

That command generates a .zip file which will containt a folder for every node specified on `instances.yml` and one folder `ca/`.

### Propagating certificate

Each instance will evaluate **node_certs_generator** (ES, Kibana and Filebeat). If it's an ES instance will generate the certificate as previously said.

In all cases such variable defines if the certificate must be copied locally (they are on same machine were the certs where generated) or execute rsync to the remote machine that contains them.

From where the certificates are copied and to where are they copied is defined by `node_certs_source` and `node_certs_destination`.

When locally, will simply copy the folder to designed destination (For example: /tmp/certs/node-1 to /etc/kibana/certs)

When remote, will execute rsync to the generator instance to sync the remote folder.

https://github.com/wazuh/wazuh-ansible/blob/2c14392e74cb3b1b85d5c3c1a7cf69e5ea36c69d/roles/elastic-stack/ansible-kibana/tasks/main.yml#L45-L52

### Template configuration

In order to enable XPack, the .yml files must be configured. The templates contain the required fields to generate the configuration file with the XPack parameters included. For example, the Kibana template:

https://github.com/wazuh/wazuh-ansible/blob/2c14392e74cb3b1b85d5c3c1a7cf69e5ea36c69d/roles/elastic-stack/ansible-kibana/templates/kibana.yml.j2#L106-L114

As previously said, such parameters will be only included if `kibana_xpack_security` flag is enabled. Same applies to ES and Filebeat

### Kibana or Filebeat nodes

Previously stated changes not only apply to Elasticsearch, it's possible to deploy an instance with only a Kibana service or a Filebeat. You can find an example on the following examples section.

### Examples:

An example of deploying a **cluster of ES instances** is written on:

https://github.com/wazuh/wazuh-ansible/blob/2c14392e74cb3b1b85d5c3c1a7cf69e5ea36c69d/playbooks/wazuh-elastic_stack-distributed.yml#L3-L53

An example where we have **separated instances of Kibana, Filebeat and ES**

https://github.com/wazuh/wazuh-ansible/blob/2c14392e74cb3b1b85d5c3c1a7cf69e5ea36c69d/playbooks/wazuh-elastic_stack-distributed.yml#L3-L29
+
https://github.com/wazuh/wazuh-ansible/blob/2c14392e74cb3b1b85d5c3c1a7cf69e5ea36c69d/playbooks/wazuh-elastic_stack-distributed.yml#L56-L86

Best regards,

Jose

**Comment by manuasir** [comment]
Implemented here #199
